### ISSUE TITLE
Data may be resurrected aftre truncate by hinted handoff

### ISSUE THREAD

**Issue by gleb-cloudius** [scylladb/scylla#6096]
Installation details*
Scylla version (or git commit hash): 3159ad4484b26f424d9a4657c58f678d76db70d4

HH does not check a table's truncation time when sending or applying a hint which may cause data to be resurrected.

**Comment by piodul** [comment]
We had a discussion with @haaawk on how to solve this issue without causing races between hints and truncation, and we came up with an idea for the solution.

The main part of the idea is to make the truncation a two-step process. Both steps consist of coordinator sending a message and sending for all replies. Step 2 is only executed if step 1 is successful. So, it would be something like a two-step commit.

The first step would enable a temporary "hint quarantine" for the table being truncated. It would prevent a number of things from happening during that period:

- New hints for subject table would not be written to hints commitlog,
- While replaying hints from the commitlog, hints for subject table would not be sent,
- If a hint mutation for subject table is received, it is not applied locally.

The second step would delete table data on local nodes, and revoke the hint quarantine.

The last (small but important) part of the solution is to prevent hints older than the truncation time of their table from being sent out. We would also flush hints to disk right at the beginning of the quarantine so that we do not mix hints older and newer than truncation time in one commitlog segment, and we can use hint segment creation time for the purpose of the mentioned check.

We also consider adding a timeout for the hint quarantine.

@gleb @vladzcloudius what do you think about this solution?

**Comment by vladzcloudius** [comment]
@piodul Looks good but there should be a very specific sequence of steps which is kinda speared in the description above.
On reception of a TRUNCATE request:
 1) stop generating any hints and/or applying any mutations to the table in question.
 2) Flush all pending hints to the disk.
 3) Remember the current time point as a cut off time (let's call it time X).
 4) Delete all sstables.
 5) On success store time X as a time of the TRUNCATE operation.
     Otherwise return a failure and inform the user that the data for the table may be in an inconsistent state. (In this case the user is expected to retry the TRUNCATE operation until it succeeds).

When replaying hints:
 1) Read the last TRUNCATE operation time for the table the hint is destined to (let's call it Y).
 2) If the write time of the hint segment is older than Y - drop this hint and move on.

We already read the segment write time for GC checks.
So, the only thing that is missing in the write path is reading that last TRUNCATE time. I hope it's not going to translate into disk I/O.
If it is then you may want to cache it and flush that cache every time you get a new TRUNCATE operation (for simplicity).

**Comment by piodul** [comment]
@gleb-cloudius ping

**Comment by gleb-cloudius** [comment]
Why can't we do the same as we do for checking that hint are not older than gc grace period: replay a hint only if  the oldest hint in a file is younger that truncation time?

**Comment by haaawk** [comment]
If I understood correctly, we're mixing hints for different tables in one file, no? @gleb-cloudius

**Comment by gleb-cloudius** [comment]
Yes, but the same is true for gc period. It is per table as well. You do the check for each mutation while replaying a segment. Drop those that are too old.

**Comment by haaawk** [comment]
I agree that we can check each mutation while replaying a segment. We just can't check the whole file as a single unit.

**Comment by piodul** [comment]
@gleb-cloudius

Adding the check you are talking about is also a part of the solution, but it won't suffice.

Currently, truncate operation is now implemented in such way that coordinator sends a request to all nodes to remove their local data, and waits until all nodes reply. Truncate requests from coordinator may arrive to each node at a different time and it may race with hint sending. An example:

- A is in the process of sending hints to B,
- B receives command from coordinator and truncates local data,
- A sends some hints to B,
- A receives command from coordinator, truncates local data and stops sending hints because of that.

In this scenario, B will have some data replayed from hints that were generated before truncation. B would need to have a way to determine if the hints it receives were generated before, or after the truncation.

A two-step process (first stop hints for the table, then actually truncate) would solve this race.

**Comment by gleb-cloudius** [comment]
@haaawk I did not suggest that, only use file's modification time as a basis for time calculation like we do for gc.

**Comment by haaawk** [comment]
> @haaawk I did not suggest that, only use file's modification time as a basis for time calculation like we do for gc.

I see. No I understand. Makes total sense to not store modification time per mutation.

**Comment by piodul** [comment]
@haaawk As far as the file modification time goes, we would need to flush currently accumulated hints just before we truncate locally. If we do that, we will not mix hints generated before/after the truncation in one file, and we can use the file modification time to check if the hint is OK to be sent.

**Comment by gleb-cloudius** [comment]
@piodul about the race I think it can be solved by not accepting any hints for truncation timeout.

**Comment by piodul** [comment]
@gleb-cloudius Oh, that's a good idea! If we reject hints with an error:

- hints generated before the truncation will be retried until they start being ignored due to them being older than the truncation,
- hints generated after the truncation will be retried until they start being accepted by the destination.

This way we will neither replay the old hints after truncation, nor lose new hints generated right after truncation. This will also be much simpler in terms of implementation complexity than the two-phased solution.

The only issue I see with this is that it will block sending hints for other tables as well - if writing one hint fails, then hints manager will keep retrying it and won't move to the next commitlog segment until that hint succeeds. I think we shouldn't worry about it, though, because the truncation timeout is not very long by default (1 minute) and it is probably better to wait than e.g. drop those hints instead.

**Comment by vladzcloudius** [comment]
> @piodul about the race I think it can be solved by not accepting any hints for truncation timeout.

@piodul
After the TRUNCATE handling is complete hints generated before the truncation should be simply dropped on the sender side. No need to do anything special about it - just like we do for gc_grace case. I described how to do it in detail before.

The node that received the TRUNCATE command should stop applying any hints and/or regular mutations white it processes the TRUNCATE. This is also in the algorithm I posted before. 

No matter how long you wait on the hint receiver side until you start accepting hints for the truncated table it's not going to help to avoid the resurrection. - Hints may simply be not sent due to network partitioning and partitionoing may take much longer than a truncate timeout.

Unless we stop hints sending cluster wide and sync the "time X" among all nodes before continuing there's always going to be an issue about hints generated between the time the first node received the TRUNCATE and the time the last one does simply due to the fact that different nodes will consider different cut-off time.

That's exactly why we say that regular READ and WRITE operations should be stopped during the TRUNCATE and if they are not - resurrection may happen.

If we want to keep on using that broken-by-design TRUNCATE implementation we need to at least stop hints sending which are technically WRITEs - and only then continue to the actual TRUNCATE.

The best solution that would resolve all our "resurrection problems" would be to implement the TRUNCATE as a range tombstone that shadows the whole token ring and is applied with CL=ALL (in order to keep the same TRUNCATE CQL semantics as today).

In addition we could have gone even further and allow different CLs for a TRUNCATE, in the same way as for regular WRITEs.

**Comment by piodul** [comment]
@vladzcloudius

> The node that received the TRUNCATE command should stop applying any hints and/or regular mutations white it processes the TRUNCATE. This is also in the algorithm I posted before.

Oh, I see. I missed that detail before. It makes sense to reject those writes during truncation.

> The best solution that would resolve all our "resurrection problems" would be to implement the TRUNCATE as a range tombstone that shadows the whole token ring and is applied with CL=ALL (in order to keep the same TRUNCATE CQL semantics as today).

If TRUNCATE was implemented as a kind of a range tombstone, it would have different semantics - after truncating you wouldn't be able to write data with timestamps older than the truncate tombstone. Current implementation does not have this limitation.

----

If there are no objections, I'll work on the solution described by @vladzcloudius.

@haaawk @gleb-cloudius

**Comment by vladzcloudius** [comment]
@piodul 

> If TRUNCATE was implemented as a kind of a range tombstone, it would have different semantics - after truncating you wouldn't be able to write data with timestamps older than the truncate tombstone. Current implementation does not have this limitation.

Actually, this is a good example of why the "tombstone TRUNCATE" would make much more sense than what we have today:
  * You *would* be able to write with (user) timestamps before the truncation time just fine. However these writes would be shadowed (at least as long as this tombstone is not evicted). 

IMO this is a much more sane semantics than what we have today - seeing data with a timestamp prior the TRUNCATE is confusing at best.

**Comment by kostja** [comment]
I suggest the issue is fixed by applying MVCC principles to schema versioning. Each mutation should have a schema versioning.  The fix could be in switching to the new format of schema hash, which doesn't include some volatile items which do not impact layout, but explicitly includes hash of the column mapping and truncation generation. When applying a mutation with an older truncation generation we will simply skip it. The truncate itself does not need two phases this way.

**Comment by avikivity** [comment]
Good idea. We already have most of the machinery in place. We'd need to add the truncation timestamp to the schema. and adjust mutation_partition::upgrade() to check if timestamps cross the truncate boundary.

**Comment by avikivity** [comment]
In fact we have so much of the machinery in place that it can be quite a small patch.

**Comment by avikivity** [comment]
> When applying a mutation with an older truncation generation we will simply skip it. The truncate itself does not need two phases this way.

Note that a mutation can include multiple timestamps (and often does). So we need to filter out cells that have timestamps from before the truncation.
